Multi-proc build fails with ArgumentException if the CreateItem task is used

### Steps to reproduce

1.proj
```xml
<Project DefaultTargets="Build">
    <Target Name="Build">
        <MSBuild Projects="2.proj" Targets="Build" />
    </Target>
</Project>
```
2.proj
```xml
<Project DefaultTargets="Build">
    <Target Name='Build' Returns="@(x)">
        <CreateItem Include='$(programfiles)\reference assemblies\**\*.dll;'>
            <Output TaskParameter='Include' ItemName='x' />
        </CreateItem>
    </Target>
</Project>
```

build with
```
set MSBUILDNOINPROCNODE=1
set MSBUILDTARGETRESULTCOMPRESSIONTHRESHOLD=0

msbuild.exe 1.proj /clp:ShowCommandLine /m:1 /verbosity:n /nr:false /bl /fl /flp:v=diagnostic
```
### Expected behavior
Build succeeds.

### Actual behavior
```
error MSB4018: The "MSBuild" task failed unexpectedly.
error MSB4018: This is an unhandled exception from a task -- PLEASE OPEN A BUG AGAINST THE TASK OWNER.
error MSB4018: System.ArgumentException: The "RecursiveDir" name is reserved, and cannot be used.
error MSB4018:    at Microsoft.Build.Shared.ErrorUtilities.ThrowArgument(Exception innerException, String resourceName, Object[] args) in C:\src\msbuild\src\Shared\ErrorUtilities.cs:line 482
error MSB4018:    at Microsoft.Build.Shared.ErrorUtilities.VerifyThrowArgument(Boolean condition, Exception innerException, String resourceName, Object arg0) in C:\src\msbuild\src\Shared\ErrorUtilities.cs:line 614
error MSB4018:    at Microsoft.Build.Shared.ErrorUtilities.VerifyThrowArgument(Boolean condition, String resourceName, Object arg0) in C:\src\msbuild\src\Shared\ErrorUtilities.cs:line 515
error MSB4018:    at Microsoft.Build.Execution.ProjectMetadataInstance..ctor(String name, String escapedValue, Boolean allowItemSpecModifiers) in C:\src\msbuild\src\Build\Instance\ProjectMetadataInstance.cs:line 72
error MSB4018:    at Microsoft.Build.Execution.ProjectMetadataInstance..ctor(String name, String escapedValue) in C:\src\msbuild\src\Build\Instance\ProjectMetadataInstance.cs:line 45
error MSB4018:    at Microsoft.Build.Execution.ProjectItemInstance.TaskItem.TranslateWithInterning(ITranslator translator, LookasideStringInterner interner) in C:\src\msbuild\src\Build\Instance\ProjectItemInstance.cs:line 1620
error MSB4018:    at Microsoft.Build.Execution.ProjectItemInstance.TaskItem..ctor(ITranslator translator, LookasideStringInterner interner) in C:\src\msbuild\src\Build\Instance\ProjectItemInstance.cs:line 859
error MSB4018:    at Microsoft.Build.Execution.ProjectItemInstance.TaskItem.FactoryForDeserialization(ITranslator translator, LookasideStringInterner interner) in C:\src\msbuild\src\Build\Instance\ProjectItemInstance.cs:line 1581
error MSB4018:    at Microsoft.Build.Execution.TargetResult.ItemsStore.DecompressItems() in C:\src\msbuild\src\Build\BackEnd\Shared\TargetResult.cs:line 511
error MSB4018:    at Microsoft.Build.Execution.TargetResult.ItemsStore.get_Items() in C:\src\msbuild\src\Build\BackEnd\Shared\TargetResult.cs:line 426
error MSB4018:    at Microsoft.Build.Execution.TargetResult.get_Items() in C:\src\msbuild\src\Build\BackEnd\Shared\TargetResult.cs:line 96
error MSB4018:    at Microsoft.Build.BackEnd.TaskHost.<BuildProjectFilesInParallelAsync>d__58.MoveNext() in C:\src\msbuild\src\Build\BackEnd\Components\RequestBuilder\TaskHost.cs:line 942
```

### Environment data
`msbuild /version` output:

```
Microsoft (R) Build Engine version 16.5.0+d4cbfca49 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

16.5.0.12403
```

as well as current master (9f3e4e72).

--- a/src/Build/Instance/ProjectItemInstance.cs
+++ b/src/Build/Instance/ProjectItemInstance.cs
@@ -1617,7 +1617,7 @@ namespace Microsoft.Build.Execution
                         {
                             int key = translator.Reader.ReadInt32();
                             int value = translator.Reader.ReadInt32();
-                            _directMetadata.Set(new ProjectMetadataInstance(interner.GetString(key), interner.GetString(value)));
+                            _directMetadata.Set(new ProjectMetadataInstance(interner.GetString(key), interner.GetString(value), true /* allow built-in metadata names */));
                         }
                     }
                 }

--- a/src/Build/Instance/ProjectItemInstance.cs
+++ b/src/Build/Instance/ProjectItemInstance.cs
@@ -1617,7 +1617,7 @@ namespace Microsoft.Build.Execution
                         {
                             int key = translator.Reader.ReadInt32();
                             int value = translator.Reader.ReadInt32();
-                            _directMetadata.Set(new ProjectMetadataInstance(interner.GetString(key), interner.GetString(value), true /* allow built-in metadata names */));
+                            _directMetadata.Set(new ProjectMetadataInstance(interner.GetString(key), interner.GetString(value), allowItemSpecModifiers: true));
                         }
                     }
                 }

--- a/src/Tasks.UnitTests/CreateItem_Tests.cs
+++ b/src/Tasks.UnitTests/CreateItem_Tests.cs
@@ -1,12 +1,16 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System;
 using System.IO;
+using System.Collections.Generic;
+using Microsoft.Build.Execution;
 using Microsoft.Build.Framework;
 using Microsoft.Build.Tasks;
 using Microsoft.Build.Utilities;
 using Xunit;
 using Xunit.Abstractions;
+using Shouldly;
 
 namespace Microsoft.Build.UnitTests
 {
@@ -154,6 +158,45 @@ namespace Microsoft.Build.UnitTests
 
             ObjectModelHelpers.AssertFileExistsInTempProjectDirectory(Path.Combine("Destination", "Foo.txt"));
             ObjectModelHelpers.AssertFileExistsInTempProjectDirectory(Path.Combine("Destination", "Subdir", "Bar.txt"));
+        }
+
+        /// <summary>
+        /// Using the CreateItem task to expand wildcards and verifying that the RecursiveDir metadatum is successfully
+        /// serialized/deserialized cross process.
+        /// </summary>
+        [Fact]
+        public void RecursiveDirOutOfProc()
+        {
+            ObjectModelHelpers.DeleteTempProjectDirectory();
+
+            string projectFileFullPath = ObjectModelHelpers.CreateFileInTempProjectDirectory("Myapp.proj", @"
+                <Project ToolsVersion=`msbuilddefaulttoolsversion` xmlns=`msbuildnamespace`>
+                  <Target Name =`Repro` Returns=`@(Text)`>
+                    <CreateItem Include=`**\*.txt`>
+                      <Output TaskParameter=`Include` ItemName=`Text`/>
+                    </CreateItem>
+                  </Target>
+                </Project>
+                ");
+
+            ObjectModelHelpers.CreateFileInTempProjectDirectory(Path.Combine("Subdir", "Bar.txt"), "bar");
+
+            string originalCompressionThresholdVariable = Environment.GetEnvironmentVariable("MSBUILDTARGETRESULTCOMPRESSIONTHRESHOLD");
+            Environment.SetEnvironmentVariable("MSBUILDTARGETRESULTCOMPRESSIONTHRESHOLD", "0");
+            try
+            {
+                BuildRequestData data = new BuildRequestData(projectFileFullPath, new Dictionary<string, string>(), null, new string[] { "Repro" }, null);
+                BuildParameters parameters = new BuildParameters();
+                parameters.DisableInProcNode = true;
+                parameters.Loggers = new ILogger[] { new MockLogger(_testOutput) };
+                BuildResult result = BuildManager.DefaultBuildManager.Build(parameters, data);
+                result.OverallResult.ShouldBe(BuildResultCode.Success);
+                result.ResultsByTarget["Repro"].Items[0].GetMetadata("RecursiveDir").ShouldBe("Subdir\\");
+            }
+            finally
+            {
+                Environment.SetEnvironmentVariable("MSBUILDTARGETRESULTCOMPRESSIONTHRESHOLD", originalCompressionThresholdVariable);
+            }
         }
 
         /// <summary>

--- a/src/Tasks.UnitTests/CreateItem_Tests.cs
+++ b/src/Tasks.UnitTests/CreateItem_Tests.cs
@@ -186,9 +186,12 @@ namespace Microsoft.Build.UnitTests
             env.SetEnvironmentVariable("MSBUILDTARGETRESULTCOMPRESSIONTHRESHOLD", "0");
 
             BuildRequestData data = new BuildRequestData(projectFileFullPath, new Dictionary<string, string>(), null, new string[] { "Repro" }, null);
-            BuildParameters parameters = new BuildParameters();
-            parameters.DisableInProcNode = true;
-            parameters.Loggers = new ILogger[] { new MockLogger(_testOutput) };
+            BuildParameters parameters = new BuildParameters
+            {
+                DisableInProcNode = true,
+                EnableNodeReuse = false,
+                Loggers = new ILogger[] { new MockLogger(_testOutput) },
+            };
             BuildResult result = BuildManager.DefaultBuildManager.Build(parameters, data);
             result.OverallResult.ShouldBe(BuildResultCode.Success);
             result.ResultsByTarget["Repro"].Items[0].GetMetadata("RecursiveDir").ShouldBe("Subdir" + Path.DirectorySeparatorChar);

--- a/src/Tasks.UnitTests/CreateItem_Tests.cs
+++ b/src/Tasks.UnitTests/CreateItem_Tests.cs
@@ -191,7 +191,7 @@ namespace Microsoft.Build.UnitTests
                 parameters.Loggers = new ILogger[] { new MockLogger(_testOutput) };
                 BuildResult result = BuildManager.DefaultBuildManager.Build(parameters, data);
                 result.OverallResult.ShouldBe(BuildResultCode.Success);
-                result.ResultsByTarget["Repro"].Items[0].GetMetadata("RecursiveDir").ShouldBe("Subdir\\");
+                result.ResultsByTarget["Repro"].Items[0].GetMetadata("RecursiveDir").ShouldBe("Subdir" + Path.PathSeparator);
             }
             finally
             {

--- a/src/Tasks.UnitTests/CreateItem_Tests.cs
+++ b/src/Tasks.UnitTests/CreateItem_Tests.cs
@@ -167,6 +167,8 @@ namespace Microsoft.Build.UnitTests
         [Fact]
         public void RecursiveDirOutOfProc()
         {
+            using var env = TestEnvironment.Create(_testOutput);
+
             ObjectModelHelpers.DeleteTempProjectDirectory();
 
             string projectFileFullPath = ObjectModelHelpers.CreateFileInTempProjectDirectory("Myapp.proj", @"
@@ -181,22 +183,15 @@ namespace Microsoft.Build.UnitTests
 
             ObjectModelHelpers.CreateFileInTempProjectDirectory(Path.Combine("Subdir", "Bar.txt"), "bar");
 
-            string originalCompressionThresholdVariable = Environment.GetEnvironmentVariable("MSBUILDTARGETRESULTCOMPRESSIONTHRESHOLD");
-            Environment.SetEnvironmentVariable("MSBUILDTARGETRESULTCOMPRESSIONTHRESHOLD", "0");
-            try
-            {
-                BuildRequestData data = new BuildRequestData(projectFileFullPath, new Dictionary<string, string>(), null, new string[] { "Repro" }, null);
-                BuildParameters parameters = new BuildParameters();
-                parameters.DisableInProcNode = true;
-                parameters.Loggers = new ILogger[] { new MockLogger(_testOutput) };
-                BuildResult result = BuildManager.DefaultBuildManager.Build(parameters, data);
-                result.OverallResult.ShouldBe(BuildResultCode.Success);
-                result.ResultsByTarget["Repro"].Items[0].GetMetadata("RecursiveDir").ShouldBe("Subdir" + Path.DirectorySeparatorChar);
-            }
-            finally
-            {
-                Environment.SetEnvironmentVariable("MSBUILDTARGETRESULTCOMPRESSIONTHRESHOLD", originalCompressionThresholdVariable);
-            }
+            env.SetEnvironmentVariable("MSBUILDTARGETRESULTCOMPRESSIONTHRESHOLD", "0");
+
+            BuildRequestData data = new BuildRequestData(projectFileFullPath, new Dictionary<string, string>(), null, new string[] { "Repro" }, null);
+            BuildParameters parameters = new BuildParameters();
+            parameters.DisableInProcNode = true;
+            parameters.Loggers = new ILogger[] { new MockLogger(_testOutput) };
+            BuildResult result = BuildManager.DefaultBuildManager.Build(parameters, data);
+            result.OverallResult.ShouldBe(BuildResultCode.Success);
+            result.ResultsByTarget["Repro"].Items[0].GetMetadata("RecursiveDir").ShouldBe("Subdir" + Path.DirectorySeparatorChar);
         }
 
         /// <summary>

--- a/src/Tasks.UnitTests/CreateItem_Tests.cs
+++ b/src/Tasks.UnitTests/CreateItem_Tests.cs
@@ -191,7 +191,7 @@ namespace Microsoft.Build.UnitTests
                 parameters.Loggers = new ILogger[] { new MockLogger(_testOutput) };
                 BuildResult result = BuildManager.DefaultBuildManager.Build(parameters, data);
                 result.OverallResult.ShouldBe(BuildResultCode.Success);
-                result.ResultsByTarget["Repro"].Items[0].GetMetadata("RecursiveDir").ShouldBe("Subdir" + Path.PathSeparator);
+                result.ResultsByTarget["Repro"].Items[0].GetMetadata("RecursiveDir").ShouldBe("Subdir" + Path.DirectorySeparatorChar);
             }
             finally
             {

